### ISSUE TITLE
perf: investigate deletion-only compactions

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/pebble#720]
We have "normal" compactions which take a set of input sstables and rewrite them. We have "move" compactions which take a set of input sstables and move them to a new level. Should we have deletion-only compactions which take a set of sstables and delete them? 

The idea behind a deletion-only compaction is to delete one or more sstables that are completely covered by a range tombstone at a higher level and for which there is not a snapshot pinning some of their keys in existence. This would be a bit like the effect of the RocksDB [`DeleteFilesInRange`](https://github.com/facebook/rocksdb/blob/master/include/rocksdb/convenience.h#L463-L470) operation, but safe with regards to the normal operation semantics. This idea has come up in the past, and the recent work on incorporating range tombstones in the compaction heuristic (#707) has brought it up again. #707 is already determining that an sstable is completely covered by a range tombstone. If that occurs, we could mark a flag that all of the keys in an sstable are deleted. We then simply have to wait for a time when the range of seqnums in the sstable are in the same snapshot stripe as the covering range tombstone.

The expected result of this optimization is to make `DROP/TRUNCATE TABLE` in CRDB much more efficient.

**Comment by jbowens** [comment]
I used `pebble bench compact` to run an existing clearrange workload that I had collected a while back, and I printed out the count and size of tables completely contained by a range deletion in `estimateSizeBeneath`: 

```
Range tombstone [bd898800-bd89f90130e3b900) from L0.004373 can drop 1225 files reclaiming 20 G bytes.
Range tombstone [bd89f90130e3b900-be00) from L0.004374 can drop 2792 files reclaiming 45 G bytes.
Range tombstone [bd898800-be00) from L3.004375 can drop 4017 files reclaiming 64 G bytes.
Range tombstone [bd898800-be00) from L4.004376 can drop 4017 files reclaiming 64 G bytes.
Range tombstone [bd898800-be00) from L5.004378 can drop 3605 files reclaiming 58 G bytes.
```

Seems pretty promising. I'll try to prototype something

**Comment by petermattis** [comment]
> Seems pretty promising. I'll try to prototype something

Wow, those numbers are larger than I was expecting. As mentioned in person, the CRDB compactor queue might be having an outsize impact on the scope of these range tombstones. We may need to merge adjacent range tombstones to see the same effect when the compactor queue is gone. Regardless, I'm eager to see what a prototype reveals.

**Comment by jbowens** [comment]
Looking more at the compactor queue code, if I understand it correctly, it looks like it merges the intervals that it will eventually call `(*pebble.DB).Compact` with, but not the actual range tombstones it lays down.

Also, the experiment I did up above _did_ merge abutting range tombstones within the same file (using the table stats collector logic). The numbers might've been so large because in the clearrange workload most the range tombstones laid down were contiguous and flushed into just a couple L0 sstables. If that's the case, it does seem promising.

**Comment by petermattis** [comment]
> Looking more at the compactor queue code, if I understand it correctly, it looks like it merges the intervals that it will eventually call (*pebble.DB).Compact with, but not the actual range tombstones it lays down.

Yep, the compactor queue does merge adjacent intervals.

> Also, the experiment I did up above did merge abutting range tombstones within the same file (using the table stats collector logic). The numbers might've been so large because in the clearrange workload most the range tombstones laid down were contiguous and flushed into just a couple L0 sstables. If that's the case, it does seem promising.

The `clearrange` workload might be better named `drop-large-table`. I think this is a worthwhile workload to optimize for. Reclaiming disk space when a table is dropped takes significantly longer and requires significantly more work in CRDB than in legacy databases where such an operation is the removal of a set of files. Deletion-only compactions would move us much closer to that level of performance.

**Comment by jbowens** [comment]
I think the simplest variant of a delete-only compaction would piggy back off the table stats collector. The stats collector already merges abutting range tombstones within the same sstable and recognizes when a sstable is completely contained within one of these merged tombstones. It could accumulate these as 'delete table hints' storing the highest sequence number of the merged tombstones and the lowest sequence number of the fully-contained files. These hints could be stored in a slice on `d.mu.compact.deleteTableHints`.

The `maybeScheduleCompaction` method would check each hint to see if the highest tombstone sequence number and the lowest file sequence number are contained in the same snapshot stripe. If they are, it would recheck the current version to find the files to delete (in case they were compacted into new eligible files or into no longer eligible files). If it finds files, it marks those files as compacting and schedules a delete-only compaction of them.

There are a few limitations with that approach that we could address:

**Covered files might be split across several snapshots.**

 Waiting for _all_ the files to be in the same stripe as the tombstones could unnecessarily prevent a delete-only compaction of _some_ of the files. One possible solution is to store a sequence number per overlapping stripe in the hint. As soon as any of the sequence numbers appear in the same stripe as the tombstones, `maybeScheduleCompaction` checks for a delete-only compaction.

**Range tombstones might be contiguous but split across files.**

The grandparent limit in compaction might even make this pretty common when a tombstone makes it to levels L1+. We could try to merge delete hints from multiple sstables. I think this can get tricky. We don't want to continuously extend a delete-only hint in the space of sequence numbers, preventing it from being used because we extend it into new snapshot stripes. We also don't want to accumulate every discrete range tombstone as a deletion hint that doesn't contain any file completely, but _might_ at some point in the future.

**Range tombstones might still be in the memtable.**

Handling this could remove the need for the `delete_range_flush_delay` option. Whenever a range deletion is added to a memtable, we could try to merge it with other range tombstones in the same memtable and then check for files completely contained by the merged interval. If we find any, we could add a delete-only compaction hint. We might allow that hint to be broadened whenever a new range tombstone is added to the same memtable. This wouldn't handle ingested tombstones, which we'd need to handle separately or as part of the table stats collector. I'm not sure if ingesting broad range tombstones is common in actual CockroachDB workloads. I might try to measure this with the clearrange workload.

----

From what I've seen so far in the clearrange workload so far, I suspect we could get pretty far with just the stats collector hints and the `delete_range_flush_delay` option or with a variant of the memtable approach. Dropping a table produces a lot of contiguous range tombstones in a short time span, so they tend to end up in the same memtable or L0 sstable.

**Comment by petermattis** [comment]
I'm thumbs-up on starting with the stats collector hints approach.

> I think the simplest variant of a delete-only compaction would piggy back off the table stats collector. The stats collector already merges abutting range tombstones within the same sstable and recognizes when a sstable is completely contained within one of these merged tombstones. It could accumulate these as 'delete table hints' storing the highest sequence number of the merged tombstones and the lowest sequence number of the fully-contained files. These hints could be stored in a slice on d.mu.compact.deleteTableHints.

The hint also contain the file numbers of the sstables that are completely covered, right?

> Range tombstones might be contiguous but split across files.

I think this may be a larger concern now that L0-sublevels and flush-split-bytes has landed.

> Range tombstones might still be in the memtable.
> 
> ... Whenever a range deletion is added to a memtable, we could try to merge it with other range tombstones in the same memtable and then check for files completely contained by the merged interval.

The specific proposal could make adding range tombstones quite a bit more expensive. Rather than doing the check every time a range tombstone is added, I'd suggest doing something where the addition of a range tombstone starts a timer for some short duration in the future (e.g. 5s). The addition of a new tombstone resets that timer and if the memtable gets rotated the timer gets reset. We'd be replacing `delete_range_flush_delay` with `delete_range_check_delay`.

**Comment by jbowens** [comment]
> The hint also contain the file numbers of the sstables that are completely covered, right?

I think it doesn't need to, because we need to check the version for the file's existence regardless. We mind as well use `(*Version).Overlaps` and re-evaluate eligibility of the files so that we can pick up eligible files that were created through a compaction in the interim.

> I think this may be a larger concern now that L0-sublevels and flush-split-bytes has landed.

Good point.

>  I'd suggest doing something where the addition of a range tombstone starts a timer for some short duration in the future (e.g. 5s). The addition of a new tombstone resets that timer and if the memtable gets rotated the timer gets reset. We'd be replacing delete_range_flush_delay with delete_range_check_delay.

That makes sense.

**Comment by jbowens** [comment]
I ran a local modified `clearrange` roachtest using a prototype, and it looks promising. I'm going to try running it with an unmodified cloud `clearrange` roachtest, once I figure out how to build a linux version of cockroach w/ my local pebble branch.

![Screen Shot 2020-06-26 at 4 51 52 PM](https://user-images.githubusercontent.com/867352/85900425-d2246280-b7cd-11ea-8546-6abe175167a7.png)
![Screen Shot 2020-06-26 at 4 52 04 PM](https://user-images.githubusercontent.com/867352/85900434-d5b7e980-b7cd-11ea-9ef9-127eecbb2487.png)

From the logs, it looks like there was one big delete-only compaction that deleted 5.8 G from L5 and 58 G from L6. Actually deleting the files in `doDeleteObsoleteFiles` took ~5s, but the `CompactionInfo` recorded the compaction as taking 0.0s, since all it did was build the version edit.

**Comment by petermattis** [comment]
> From the logs, it looks like there was one big delete-only compaction that deleted 5.8 G from L5 and 58 G from L6. Actually deleting the files in doDeleteObsoleteFiles took ~5s, but the CompactionInfo recorded the compaction as taking 0.0s, since all it did was build the version edit.

Very cool! We may have to throttle the sstable deletions. That's a nice problem to have. When this gets developed further, it would be good to provide the delta in reclaimed space numerically rather than graphically. I had to go back to https://github.com/cockroachdb/cockroach/pull/50508 to see how much of an improvement this is. Looks like ~10x faster, but that's a bit hard to see given the different time scales.

**Comment by jbowens** [comment]
@petermattis here's a description of the bug that I mentioned in standup.

My original implementation used compaction hints that recorded the LSM level, user key range, the tombstones' sequence number ranges and the minimum sequence number in completely covered files. It didn't record the actual files that were covered, because I figured that we would need to rescan the version at compaction anyways.

The metamorphic tests found a failure that looks like:

* A file 000100.sst containing keys with sequence numbers 100-200 is flushed into L0 with a range tombstone `a.RANGEDEL.100:z`.
* The stats collector starts loading stats for 000100.sst with the current version.
* A new, ordinary compaction starts, compacting 000100.sst into L6.
* The table stats collector loads table stats for 000100.sst and finds files completely contained by the range tombstone, including a file in L6 with zeroed sequence numbers. It records a hint. `L0 a-z seqnums(tombstones=100-100, file-smallest=0)`.
* The ordinary compaction finishes compacting 000100.sst into L6, creating a new file 000200.sst with zeroed sequence numbers. The file includes keys from 000100.sst that originally had sequence numbers 101-200.
* `maybeScheduleCompaction` checks for a delete-only compaction, and finds that no snapshots prevent the hint from being resolved. It sees that the file L6 000200.sst falls within the hint's user key range, has sequence numbers less than the tombstone hint and deletes it, dropping keys that should not have been dropped.

**Comment by jbowens** [comment]
I'm not confident whether this problem can only manifest if the range tombstone's file itself is compacted. I've been trying to come up with a scenario that doesn't involve the range tombstone's file being compacted, but I'm failing.

If it only occurs when the tombstone itself is compacted, we could check for the tombstone's file in the version before running a delete-only compaction.

**Comment by jbowens** [comment]
> I'm not confident whether this problem can only manifest if the range tombstone's file itself is compacted. I've been trying to come up with a scenario that doesn't involve the range tombstone's file being compacted, but I'm failing.
>
> If it only occurs when the tombstone itself is compacted, we could check for the tombstone's file in the version before running a delete-only compaction.

The more I think about this, the more I confident I become about it. If the tombstone's file hasn't been compacted, that means that no keys within the tombstone's range w/ higher sequence numbers may be below it in the LSM, because that would violate the sequence number invariant.

**Comment by petermattis** [comment]
Interesting scenario. Thanks for the write-up, @jbowens. Seems like it is ok if the sstable containing the range tombstone has been compacted, as long as the compaction didn't allow zero seqnum. (See `compaction.allowZeroSeqNum()`). If the range tombstone was part of a compaction that allowed zeroing of seqnum, it means there are no sstables below it anyways.

Do you have a thought for how to handle this? Perhaps you can remove compaction hints when certain compactions occur.

**Comment by jbowens** [comment]
> Seems like it is ok if the sstable containing the range tombstone has been compacted, as long as the compaction didn't allow zero seqnum.

Agreed, though this seems tricky to reliably detect, because the compaction of the sstable containing the range tombstone might happen before the table stats collector appends the compaction hint.

> Do you have a thought for how to handle this?

I've been struggling to come up with a scheme that reliably handles this that doesn't skip hints whose sstable was compacted OR reduces the asynchrony of the table stats collector. If the table stats were loaded from the compaction job like we talked about before, I think we would be able to clear compaction hints when conflicting zero-seqnum compactions occur. Not sure yet which is preferable.

**Comment by jbowens** [comment]
> If the table stats were loaded from the compaction job like we talked about before, I think we would be able to clear compaction hints when conflicting zero-seqnum compactions occur.

Whoops, I don't think this is true unless we completely eliminate the backfilling of stats that existed at Open, because a backfilled tables' stats can always be loaded after the table has already been compacted.

It might be sufficient to clear conflicting hints at the end of a compaction _and_ check that the current version still includes the tombstone's sstable when _appending_ the compaction hint, but I'm a little worried about this approach being susceptible to subtle races.

**Comment by jbowens** [comment]
**with deletion-only compactions**
![Screen Shot 2020-07-06 at 1 44 55 PM](https://user-images.githubusercontent.com/867352/86624169-dc461e00-bf90-11ea-8108-874141087fab.png)
![Screen Shot 2020-07-06 at 1 54 13 PM](https://user-images.githubusercontent.com/867352/86624199-e9630d00-bf90-11ea-84b6-a45a68dff56e.png)

**without deletion-only compactions**
![Screen Shot 2020-07-06 at 6 12 15 PM](https://user-images.githubusercontent.com/867352/86655129-7a97ab00-bfb4-11ea-89ba-f8b1efa939e9.png)
![Screen Shot 2020-07-06 at 6 13 38 PM](https://user-images.githubusercontent.com/867352/86655145-7d929b80-bfb4-11ea-874f-5eb8c2a1e668.png)

I measured the time taken from the gc job status switching to `DELETING` until the cluster capacity hit 0.04 TiB for both clusters. With deletion-only compactions, it took 79s. Without it took 484s, so it provided somewhere around a 6x speedup of the deletion.

With deletion-only compactions, there's still a bit of a long-tail of ordinary compactions for sstables that did not completely fall within a hint. You can see it in the disk-write ops graph that continue well after the bulk of the data has been dropped.

**Comment by petermattis** [comment]
> It might be sufficient to clear conflicting hints at the end of a compaction and check that the current version still includes the tombstone's sstable when appending the compaction hint, but I'm a little worried about this approach being susceptible to subtle races.

I don't have anything intelligent to say about this right now. I'll try to give this a close look later in the week.

> With deletion-only compactions, it took 79s. Without it took 484s, so it provided somewhere around a 6x speedup of the deletion.

Very nice!

> With deletion-only compactions, there's still a bit of a long-tail of ordinary compactions for sstables that did not completely fall within a hint. You can see it in the disk-write ops graph that continue well after the bulk of the data has been dropped.

This is interesting, though expected now that I see it. Deletion-only compactions shouldn't reduce bytes written, only bytes read. Eye-balling the graphs, that seems to be what is happening.

**Comment by jbowens** [comment]
Merged in #784.
